### ISSUE TITLE
Base#toJSON fails when global `Object` is substituted

### ISSUE THREAD

**Issue by brettz9** [w3c/webidl2.js#426]
I apologize in advance for my lack of in-depth research into the parser... And I can work around this issue, but I thought I'd at least report what I am running into in my own environment.

I am testing [IndexedDBShim](https://github.com/axemclion/IndexedDBShim/) in Node using web-platform-tests (which uses webidl2).

I ran across an error which shows up when `Base#toJSON` gets called and which I can readily fix by changing the following (also [here](https://github.com/web-platform-tests/wpt/blob/master/resources/webidl2/lib/webidl2.js#L1067) in web-platform-tests):

https://github.com/w3c/webidl2.js/blob/30f749b648368173bdbf0abdb251ff0c433bd97f/lib/productions/base.js#L12 

to:

```js
while (proto && proto !== Object.prototype) {
```

...as apparently I have some `Object.create()` objects getting through to this point in the script.

If this condition wouldn't occur for others or is mistaken in some way, then feel free to close the issue of course, but just reporting in case other consumers might end up needing this or if there is no harm in changing things here.

The original object passed in, btw, was the following (as printed to console):

```
Attribute {
  idlType:
   Type { extAttrs: ExtendedAttributes [], type: 'attribute-type' },
  extAttrs: ExtendedAttributes [] }
```

**Comment by saschanaz** [comment]
Your change indeed looks solid, but I want to understand what situation causes such error. Are you modifying AST objects by assigning `Object.create()` objects?

**Comment by brettz9** [comment]
I haven't investigated what the full chain might be, but the copy I am using is at https://github.com/web-platform-tests/wpt/blob/master/resources/webidl2/lib/webidl2.js#L1067 and it is specifically come from the running of this test: https://github.com/web-platform-tests/wpt/blob/master/IndexedDB/idlharness.any.js (the `WebIDLParser.js` mentioned as a dependence in the file is a copy of the `webidl2.js` just mentioned).

Note that I am injecting my own jsdom or other simulated browser globals, some of which have their prototypes created by `Object.create`.

**Comment by saschanaz** [comment]
The original WPT does not show such error so I think there somehow was some modification in AST in your code... But okay, your code is valid and looks good, I'll fix it.

**Comment by saschanaz** [comment]
Oh wait, that code can't throw as `toJSON()` is a method in `Base` which inherits from `Object`. I thought it's something `Object.create(null)` assigned to the AST but it's not. This is mysterious, which environment are you running on?

**Comment by brettz9** [comment]
The script is `npm run w3c` on [IndexedDBShim](https://github.com/axemclion/IndexedDBShim), but we've had to go through a lot of hoops to appease the testing environment, so it may be a bit hard to follow. But we are definitely not trying to manipulate AST--the most it should be is our modifying globals which perhaps the web-platform-tests are not expecting to be in a certain way.

**Comment by saschanaz** [comment]
Hmm. maybe something pollutes global scope and changes `Object` definition? What happens when you do:

```js
while (proto !== ({}).__proto__) { 
```

?

**Comment by brettz9** [comment]
that works too...

**Comment by saschanaz** [comment]
Hmm, that strongly suggests that the shim code does something for `Object` in a dangerous way. 🤔

**Comment by brettz9** [comment]
I think I've identified the source--I'm setting `Object` of the virtual machine with our Node `Object`. I can see about working around.

**Comment by brettz9** [comment]
Yes, I got away with avoiding needing to modify the file as I don't need the `Object` injection for the IDL tests (has something to do with `instanceof` checks, IIRC).

**Comment by saschanaz** [comment]
Closing for now because substituting `Object` value is already dangerous as substituting `null` value is.
